src/Doctrine/Common/State/PersistProcessor.php

```diff
@@ -38,10 +44,32 @@ public function __construct(private readonly ManagerRegistry $managerRegistry)
      */
     public function process(mixed $data, Operation $operation, array $uriVariables = [], array $context = []): mixed
     {
-        if (!$manager = $this->getManager($data)) {
+        if (
+            !\is_object($data) ||
+            !$manager = $this->managerRegistry->getManagerForClass($this->getObjectClass($data))
+        ) {
             return $data;
         }
 
+        // PUT: reset the existing object managed by Doctrine and merge data sent by the user in it
+        // This custom logic is needed because EntityManager::merge() has been deprecated and UPSERT isn't supported:
+        // https://github.com/doctrine/orm/issues/8461#issuecomment-1250233555
+        if (isset($context['read_data'], $context['resource_class']) && $operation instanceof HttpOperation && 'PATCH' !== $operation->getMethod()) {
+            $propertyNames = iterator_to_array($this->propertyNameCollectionFactory->create($context['resource_class'])->getIterator());
```

Comment:
we can avoid all this, you can retrieve identifiers by looping over `foreach $operation->getUriVariables() as $uriVariable` then `foreach $uriVariable->getIdentifiers() as $property` this way we don't use metadata factories here.

Reply:
Awesome.

Reply:
My refactoring doesn't handle the case where URI variables don't match the Doctrine properties, but I think that it's good enough for now, mainly because it's in sync with what is currently done here: https://github.com/api-platform/core/blob/main/src/Doctrine/Orm/State/ItemProvider.php#L51

At some point, we'll probably want to introduce a utility to map URI variables with internal Doctrine identifiers.